porcupine-core/src/System/TaskPipeline/PTask/Internal.hs

```diff
@@ -120,39 +127,92 @@ newtype PTask m a b = PTask
     (RunnablePTask m)
     a b)
   deriving (Category, Arrow, ArrowError SomeException, Functor, Applicative)
+  -- PTask doesn't instanciate ArrowChoice. That's intentional, even if an
+  -- instance could be automatically derived. The ArrowChoice implementation for
+  -- `AppArrow (Writer x) arr` isn't sane for PTasks, as the monoid state (the
+  -- PTask requirements) will be accumulated by (|||) in a way that's
+  -- indistiguishable from (>>>), that is to say that doesn't differentiate
+  -- VirtualFiles that _will_ be used from those that _may_ be used. Maybe in
+  -- the future we will implement ArrowChoice/ArrowPlus/ArrowZero in a saner way (it
+  -- should even be necessary if we want to implement serialization methods with
+  -- PTask themselves, and have serial method selection based on file format or
+  -- mapping metadata), but in that case it's necessary that the pipeline
+  -- configuration file reflects this "either-or" nature of VirtualFiles.
 
-flowToPTask :: Flow (AsyncA m) SomeException a b -> PTask m a b
+flowToPTask :: Flow (InnerEffect m) SomeException a b -> PTask m a b
 flowToPTask = PTask . appArrow . appArrow
 
 -- | The type of effects we can run. The reader layer is executed by 'wrap',
 -- this is why it doesn't appear in the Flow part of the 'RunnablePTask' type.
-type EffectInFlow m = AsyncA (ReaderT (PTaskReaderState m) m)
+type OuterEffect m =
+  AsyncA (ReaderT (PTaskState m) m)
+
+-- | The effects ran inside the flow have to handle some dynamic modifications
+-- of the state (for instance from task inputs) that have to be applied to each
+-- state passed to 'wrap'. We store these modifications as a stack of functions.
+type InnerEffect m =
+  AsyncA (StateT [PTaskState m -> PTaskState m] m)
 
-instance (KatipContext m) => ArrowFlow (EffectInFlow m) SomeException (PTask m) where
+instance (KatipContext m)
+      => ArrowFlow (OuterEffect m) SomeException (PTask m) where
   step' props f = flowToPTask $ step' props f
   stepIO' props f = flowToPTask $ stepIO' props f
   external f = flowToPTask $ external f
   external' props f = flowToPTask $ external' props f
   -- wrap' transmits the Reader state of the PTask down to the flow:
   wrap' props (AsyncA rdrAct) = runnableWithoutReqs $
-    withRunnableState' props $ \state input ->
-                              runReaderT (rdrAct input) state
+    withRunnableState' props $ \outerState input ->
+      runReaderT (rdrAct input) outerState
   putInStore f = flowToPTask $ putInStore f
   getFromStore f = flowToPTask $ getFromStore f
   internalManipulateStore f = flowToPTask $ internalManipulateStore f
 
+withOuterState
+  :: (Monad m, ArrowFlow (AsyncA m) ex arr)
+  => Properties a b
+  -> (t -> a -> m b)
+  -> AppArrow (Reader t) arr a b
+withOuterState props f =
+  AppArrow $ reader $ \outerState ->
+    wrap' props $ AsyncA $ \input ->
+      f outerState input
```

Comment:
Why is this called a state where it's a reader-like effect? Is the "outer" terminology here consistent with the `OuterEffect` type?

Reply:
Yeah... It could be `config` instead of `state`... But I mean that's the reader state...

Reply:
Environment?